docs/howto/architectures.rst

```diff
@@ -0,0 +1,234 @@
+
+To create a snap that will be built on ``amd64`` and built for ``amd64``, use
+one of the ``snapcraft.yaml`` snippets below.
+
+core22
+^^^^^^
+
+.. code-block:: yaml
+
+  architectures:
+    - build-on: amd64
+      build-for: amd64
+
+Building on ``amd64`` will produce one snap built for ``amd64``. Snapcraft will
+raise an error when building on another architecture.
+
+If ``build-for`` is omitted, then it will assume the value of ``build-on``. The
+following snippet snippet will produce the same result:
+
+.. code-block:: yaml
+
+  architectures:
+    - build-on: amd64
+
+The shorthand format will also produce the same result:
+
+.. code-block:: yaml
+
+  architectures:
+    - amd64
+
+core20
+^^^^^^
+
+.. code-block:: yaml
+
+  architectures:
+    - build-on: amd64
+      run-on: amd64
+
+Building on ``amd64`` will produce one snap built for ``amd64``. Snapcraft will
+not raise an error when building on another architecture. Instead, it will
+ignore the ``architectures`` keyword and build for the build-on architecture.
+
+If ``run-on`` is omitted, then it will assume the value of ``build-on``. The
+following snippet snippet will produce the same result:
+
+.. code-block:: yaml
+
+  architectures:
+    - build-on: amd64
+
+The shorthand format will also produce the same result:
+
+.. code-block:: yaml
+
+  architectures:
+    - amd64
+
+How to create a set of snaps for multiple architectures
+-------------------------------------------------------
+
+core22
+^^^^^^
+
+.. code-block:: yaml
+
+  architectures:
+    - build-on: amd64
+      build-for: amd64
+    - build-on: arm64
+      build-for: arm64
+
+Building on ``amd64`` will produce one snap for ``amd64``. Building on ``arm64``
+will produce one snap for ``arm64``. Snapcraft will raise an error when building
+on another architecture.
+
+If ``build-for`` is omitted, then it will assume the value of ``build-on``. The
+following snippet snippet will produce the same result:
+
+.. code-block:: yaml
+
+  architectures:
+    - build-on: amd64
+    - build-on: arm64
+
+The shorthand format will also produce the same result:
+
+.. code-block:: yaml
+
+  architectures:
+    - amd64
+    - arm64
+
+core20
+^^^^^^
+
+.. code-block:: yaml
+
+  architectures:
+    - build-on: amd64
+      run-on: amd64
+    - build-on: arm64
+      run-on: arm64
+
+Building on ``amd64`` will produce one snap built for ``amd64``. Building on
+``arm64`` will produce one snap built for ``arm64``. Snapcraft will not raise
+an error when building on another architecture. Instead, it will ignore the
+``architectures`` keyword and build for the build-on architecture.
+
+If ``run-on`` is omitted, then it will assume the value of ``build-on``. The
+following snippet snippet will produce the same result:
+
+.. code-block:: yaml
+
+  architectures:
+    - build-on: amd64
+    - build-on: arm64
+
+The shorthand format will also produce the same result:
+
+.. code-block:: yaml
+
+  architectures:
+    - amd64
+    - arm64
```

Comment:
Can one mix the short-hand and long-form? E.g.

```yaml
architectures:
  - amd64
  - build-on: [amd64, arm64]
    build-for: riscv64
```

Reply:
It's not supported and there's a mediocre error message for it:
```
Bad snapcraft.yaml content:
- every item must either be a string or an object for ['i386', Architecture(build_on='amd64', build_for='arm64')] (in field 'architectures')
```

I updated the documentation and mentioned this in the [issue](https://github.com/snapcore/snapcraft/issues/4361) to improve error messages.